{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "library(snpStats)\n",
    "library(TCA)\n",
    "library(pracma)\n",
    "library(matrixcalc)\n",
    "library(optparse)\n",
    "library(pbapply)\n",
    "library(tidyr)\n",
    "\n",
    "find_cis_snps_gene <-function(S_geno,G_locs,S_map,G,flank = 1000000){\n",
    "    G_locs = G_locs[G,]\n",
    "    S_map = S_map[S_map$chromosome==G_locs$CHROM,]\n",
    "    cis_snps = rownames(S_map[((S_map$position>G_locs$START-flank) & (S_map$position<G_locs$START))\n",
    "                              | ((S_map$position<G_locs$END+flank) & (S_map$position>G_locs$END)),])\n",
    "    return(S_geno[,colnames(S_geno) %in% cis_snps])\n",
    "}\n",
    "\n",
    "tca_per_gene <- function(snps,geno_locs,ge,prop,i){\n",
    "    k = ncol(prop)\n",
    "    n = nrow(prop)\n",
    "    W_norms <- rowSums(prop**2)**0.5\n",
    "    gene = rownames(ge)[i]\n",
    "    X_raw = find_cis_snps_gene(snps$genotype,geno_locs,snps$map,gene)\n",
    "    X = as(X_raw,'numeric')\n",
    "    W = prop\n",
    "    G = ge[i,]\n",
    "    \n",
    "    # find bad individual snps\n",
    "    bad_rows = array(0,ncol(X))\n",
    "    for (j in 1:ncol(X)){\n",
    "        X_ = matrix(X[,j],nrow=nrow(X),ncol=1)\n",
    "        C1_ = hadamard.prod(Reshape(Reshape(apply(W, 2, function(v) repmat(v,1,1)), n*k,1), n,k), repmat(X_, 1, k))\n",
    "        D = as.matrix(cbind(W,C1_))\n",
    "        if(qr(D)$rank<dim(D)[2]){\n",
    "            bad_rows[j] = 1\n",
    "        }\n",
    "    }\n",
    "    bad_rows = which(bad_rows==1)\n",
    "    bad_snps = colnames(X[,c(bad_rows)])                                               \n",
    "    X = X[,!colnames(X) %in% bad_snps]\n",
    "    \n",
    "    for(randoms in 1:2){\n",
    "        # remove across column colinearity\n",
    "        rnm = matrix(rnorm(ncol(X)*nrow(X),mean=0,sd=0.1),ncol=ncol(X),nrow=nrow(X))\n",
    "    \n",
    "        tryCatch({\n",
    "            tca.mdl = tca(X=G,W=W,C1=X+rnm,verbose=FALSE)\n",
    "            z = tensor(G,tca.mdl,log_file=NULL,verbose=FALSE)\n",
    "        \n",
    "            # parse gammas into matrix\n",
    "            gammas = tca.mdl$gammas_hat\n",
    "            rn = data.frame(colnames(gammas))\n",
    "            colnames(rn) = 'mixed_name'\n",
    "            rn = rn %>% separate('mixed_name',c('cell','gene'))\n",
    "            gammas = data.frame(matrix(gammas,nrow=dim(tca.mdl$mus_hat)[2],byrow=TRUE),\n",
    "                                row.names=unique(rn[,1]))\n",
    "            colnames(gammas) = unique(rn[,2])\n",
    "            print(paste('SUCC:',i,sep=''))\n",
    "            return(list('z'=do.call(rbind.data.frame,z),\n",
    "                        'mdl'=list('mus_hat'=tca.mdl$mus_hat,\n",
    "                        'sigmas_hat'=tca.mdl$sigmas_hat,\n",
    "                        'tau_hat'=tca.mdl$tau_hat,\n",
    "                        'deltas_hat'=tca.mdl$deltas_hat,\n",
    "                        'gammas_hat'=gammas)))\n",
    "        },error=function(e){\n",
    "            print(paste('gene',i,randoms,'th try',sep=' '))\n",
    "        })        \n",
    "    }\n",
    "    print(paste('FAIL:',i,sep=''))\n",
    "    return(list('z'=0,'mdl'=0))\n",
    "}\n",
    "                                                  \n",
    "# snp_path = '/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_filted_individual_ld_pruned'\n",
    "# geno_locs_path = '/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_gene_locs.csv'\n",
    "# GE_path = '/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_expression_without_rbc.csv'\n",
    "# prop_path = '/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_props.csv'\n",
    "\n",
    "main <- function(opt){   \n",
    "    # loading data\n",
    "    snps = read.plink(opt$snp_path)\n",
    "    geno_locs = read.csv(opt$geno_locs_path,row.names=1)\n",
    "    ge = read.csv(opt$gene_expression_path,row.names=1)\n",
    "    colnames(ge) = substr(colnames(ge),2,10)\n",
    "    prop = read.csv(opt$cell_type_proportion_path,row.names=1)\n",
    "    \n",
    "    # perform tca\n",
    "    start = opt$start_gene\n",
    "    end = opt$start_gene + opt$bulk_size\n",
    "    tca_result = pblapply(start:end,\n",
    "            function(x) tca_per_gene(snps,geno_locs,ge,prop,x))\n",
    "\n",
    "    z_result = lapply(1:length(tca_result),function(x) tca_result[[x]]$z)\n",
    "    model_result = lapply(1:length(tca_result),function(x) tca_result[[x]]$mdl)\n",
    "    \n",
    "    # saving result\n",
    "    if(!opt$tca_mdl_path=='NULL'){\n",
    "        dir.create(opt$tca_mdl_path,showWarnings = FALSE)\n",
    "        save(z_result,file=paste(opt$tca_mdl_path,'z_result',start,'_',end,'.RData'))\n",
    "        save(model_result,file=paste(opt$tca_mdl_path,'mdl_result',start,'_',end,'.RData'))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse_args(opt_parser): Error in getopt(spec = spec, opt = args) : \n  \"/u/home/q/qmaai/.local/share/jupyter/runtime/kernel-a3bed7fd-59cc-4af6-a8b9-390ab8e6db99.json\" is not a valid option, or does not support an argument\n\n",
     "output_type": "error",
     "traceback": [
      "Error in parse_args(opt_parser): Error in getopt(spec = spec, opt = args) : \n  \"/u/home/q/qmaai/.local/share/jupyter/runtime/kernel-a3bed7fd-59cc-4af6-a8b9-390ab8e6db99.json\" is not a valid option, or does not support an argument\n\nTraceback:\n",
      "1. parse_args(opt_parser)",
      "2. stop(opt)"
     ]
    }
   ],
   "source": [
    "                          \n",
    "option_list = list(\n",
    "    make_option(\"--snp_path\", type=\"character\",\n",
    "              default=\"/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_filted_individual_ld_pruned\", \n",
    "              help=\"snps file path\", metavar=\"character\"),\n",
    "    make_option(\"--geno_locs_path\",type=\"character\",\n",
    "               default=\"/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_gene_locs.csv\",\n",
    "               help=\"gene location file path\",metavar=\"character\"),\n",
    "    make_option(\"--gene_expression_path\", type=\"character\",\n",
    "              default=\"/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_expression_without_rbc.csv\", \n",
    "              help=\"gene expression path\", metavar=\"character\"),\n",
    "    make_option(\"--cell_type_proportion_path\", type=\"character\",\n",
    "              default=\"/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_props.csv\", \n",
    "              help=\"cell type proportion path\", metavar=\"character\"),\n",
    "    make_option(\"--start_gene_pose\",type=\"integer\",default=1,\n",
    "               help=\"the start position of the gene to analyse\",metavar=\"number\"),\n",
    "    make_option(\"--bulk_size\", type=\"integer\", default=3, \n",
    "              help=\"number of genes to analyze per bulk\", metavar=\"number\"),\n",
    "    make_option(\"--tca_mdl_path\",type=\"character\",\n",
    "                default=\"/u/home/q/qmaai/project-sriram/tca-twas/data/tca_result/\",\n",
    "               help='tca output save position',metavar=\"character\")\n",
    ");\n",
    "\n",
    "opt_parser = OptionParser(option_list=option_list);\n",
    "opt = parse_args(opt_parser);\n",
    "\n",
    "if (!interactive()){\n",
    "    main(opt);\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "snp_path = '/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_filted_individual_ld_pruned'\n",
    "geno_locs_path = '/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_gene_locs.csv'\n",
    "GE_path = '/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_expression_without_rbc.csv'\n",
    "prop_path = '/u/home/q/qmaai/project-sriram/tca-twas/data/pruned_data/dutch_props.csv'\n",
    "\n",
    "snps = read.plink(snp_path)\n",
    "geno_locs = read.csv(geno_locs_path,row.names=1)\n",
    "ge = read.csv(GE_path,row.names=1)\n",
    "colnames(ge) = substr(colnames(ge),2,10)\n",
    "prop = read.csv(prop_path,row.names=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"gene 109 1 th try\"\n",
      "[1] \"gene 109 2 th try\"\n",
      "[1] \"gene 109 3 th try\"\n",
      "[1] \"gene 109 4 th try\"\n",
      "[1] \"gene 109 5 th try\"\n",
      "[1] \"FAIL:109\"\n",
      "[1] \"gene 120 1 th try\"\n",
      "[1] \"gene 120 2 th try\"\n",
      "[1] \"gene 120 3 th try\"\n",
      "[1] \"gene 120 4 th try\"\n",
      "[1] \"gene 120 5 th try\"\n",
      "[1] \"FAIL:120\"\n",
      "[1] \"gene 132 1 th try\"\n",
      "[1] \"gene 132 2 th try\"\n",
      "[1] \"gene 132 3 th try\"\n",
      "[1] \"gene 132 4 th try\"\n",
      "[1] \"gene 132 5 th try\"\n",
      "[1] \"FAIL:132\"\n",
      "[1] \"gene 192 1 th try\"\n",
      "[1] \"gene 192 2 th try\"\n",
      "[1] \"gene 192 3 th try\"\n",
      "[1] \"gene 192 4 th try\"\n",
      "[1] \"gene 192 5 th try\"\n",
      "[1] \"FAIL:192\"\n",
      "[1] \"gene 197 1 th try\"\n",
      "[1] \"gene 197 2 th try\"\n",
      "[1] \"gene 197 3 th try\"\n",
      "[1] \"gene 197 4 th try\"\n",
      "[1] \"gene 197 5 th try\"\n",
      "[1] \"FAIL:197\"\n",
      "[1] \"gene 225 1 th try\"\n",
      "[1] \"gene 225 2 th try\"\n",
      "[1] \"gene 225 3 th try\"\n",
      "[1] \"gene 225 4 th try\"\n",
      "[1] \"gene 225 5 th try\"\n",
      "[1] \"FAIL:225\"\n"
     ]
    }
   ],
   "source": [
    "tca_result = pblapply(c(109,120,132,192,197,225),function(x) tca_per_gene(snps,geno_locs,ge,prop,x))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 3.6.0",
   "language": "R",
   "name": "ir36"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
