{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T20:54:56.756964Z",
     "start_time": "2019-09-03T20:54:50.162Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "Loading required package: foreach\n",
      "Loaded glmnet 2.0-18\n",
      "\n",
      "\n",
      "Attaching package: ‘pracma’\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expm, lu, tril, triu\n",
      "\n",
      "Loading required package: coda\n",
      "Loading required package: MASS\n",
      "##\n",
      "## Markov Chain Monte Carlo Package (MCMCpack)\n",
      "## Copyright (C) 2003-2019 Andrew D. Martin, Kevin M. Quinn, and Jong Hee Park\n",
      "##\n",
      "## Support provided by the U.S. National Science Foundation\n",
      "## (Grants SES-0350646 and SES-0350613)\n",
      "##\n",
      "\n",
      "Attaching package: ‘MCMCpack’\n",
      "\n",
      "The following object is masked from ‘package:matrixcalc’:\n",
      "\n",
      "    vech\n",
      "\n",
      "The following object is masked from ‘package:pracma’:\n",
      "\n",
      "    procrustes\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(glmnet)\n",
    "library(matrixStats)\n",
    "library(abind)\n",
    "library(pracma)\n",
    "library(matrixcalc)\n",
    "library(TCA)\n",
    "library(MCMCpack)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T21:59:32.699148Z",
     "start_time": "2019-09-03T21:59:32.540Z"
    }
   },
   "outputs": [],
   "source": [
    "summary_statistics <- function(mdl,train_X,test_X,train_c1,test_c1,\n",
    "                               train_G,test_G,train_Z,test_Z,beta,\n",
    "                               gamma_c1,gamma_c2,model=1){\n",
    "    # TCA estimation & parameters\n",
    "    Z_hat_tca = tensor(train_G,mdl,log_file=NULL,verbose=FALSE,debug=FALSE)\n",
    "    rmse_tca = t(do.call(rbind,lapply(1:length(Z_hat_tca),\n",
    "                                    function(x) sqrt(sum((Z_hat_tca[[x]]-train_Z[,x])^2)/length(train_Z[,x])))))\n",
    "    Z_hat_cor_tca = sapply(1:length(Z_hat_tca), function(x) cor(unlist(Z_hat_tca[x]),train_Z[,x]))\n",
    "    cell_type = ncol(mdl$W)\n",
    "    if(model==1){\n",
    "        tca_beta = t(matrix(mdl$gammas_hat[,grepl(\"_SNP\",colnames(mdl$gammas_hat))],ncol=cell_type))\n",
    "        beta_hat_cor_tca = lapply(1:cell_type,function(x) \n",
    "            cor(beta[x,],mdl$gammas_hat[,grepl(paste(\"Cell_type\",x,\".G\",sep=''),colnames(mdl$gammas_hat))]))        \n",
    "    }else{ # model 2 does not predict beta for snps\n",
    "        tca_beta = NULL\n",
    "        beta_hat_cor_tca = NULL\n",
    "    }\n",
    "    gamma_hat_cor_tca = cor(array(gamma_c1),\n",
    "                        mdl$gammas_hat[,grepl('male|smok|age', colnames(mdl$gammas_hat))])\n",
    "    if(model==1){\n",
    "        pred = cbind(train_X,train_c1)\n",
    "        test_pred = cbind(test_X,test_c1)        \n",
    "    }else{ # model 2 does not use train_X/test_X as C1\n",
    "        pred = train_c1\n",
    "        test_pred = test_c1\n",
    "    }\n",
    "    # TCA parameter direct estimation <----> C1 & X\n",
    "    Z_hat_train_second_eq_tca = sapply(1:cell_type,function(x) pred %*% mdl$gammas_hat[,((x-1)*dim(pred)[2]+1):(x*dim(pred)[2])])\n",
    "    Z_hat_train_second_eq_cor_tca = diag(cor(train_Z,Z_hat_train_second_eq_tca))\n",
    "    Z_hat_test_second_eq_tca = sapply(1:cell_type,function(x) test_pred %*% mdl$gammas_hat[,((x-1)*dim(test_pred)[2]+1):(x*dim(test_pred)[2])])\n",
    "    Z_hat_test_second_eq_cor_tca = diag(cor(test_Z,Z_hat_test_second_eq_tca))\n",
    "    \n",
    "    # Lasso bulk data regression <----> X only\n",
    "    glmnet.mdl.X.cv <- cv.glmnet(x=train_X,y=t(train_G),nfolds=5)\n",
    "    glmnet.mdl.X <- glmnet(x=train_X,y=t(train_G),lambda=glmnet.mdl.X.cv$lambda.min)\n",
    "    beta_full_X_bulk <- as.numeric(glmnet.mdl.X$beta)\n",
    "    # extract non zero predictors and recorrelate\n",
    "    predictors.X <- colnames(train_X)[which(beta_full_X_bulk!=0)]\n",
    "    beta_X <- as.matrix(c(glmnet.mdl.X$a0,as.matrix(glmnet.mdl.X$beta[predictors.X,])))\n",
    "    bias_one <- numeric(nrow(train_X))+1\n",
    "    G_hat_train_lasso <- cbind(bias_one,train_X[,predictors.X]) %*% beta_X\n",
    "    G_hat_train_cor_bulk <- cor(t(train_G),G_hat_train_lasso)\n",
    "    G_hat_test_lasso <- cbind(numeric(nrow(test_X))+1,test_X[,predictors.X]) %*% beta_X\n",
    "    G_hat_test_cor_bulk <- cor(t(test_G),G_hat_test_lasso)\n",
    "    \n",
    "    # cell type specific lasso\n",
    "    corrs = numeric(ncol(mdl$W))\n",
    "    corrs.real = numeric(ncol(mdl$W))\n",
    "    corrs.beta = numeric(ncol(mdl$W))\n",
    "    corrs.test.real = numeric(ncol(mdl$W))\n",
    "    dev_ratio = numeric(ncol(mdl$W))\n",
    "    rmse_lasso = numeric(ncol(mdl$W))\n",
    "    beta_full_cell = matrix(0,nrow=ncol(mdl$W),ncol=dim(train_X)[2])\n",
    "    Z_hat_test_lasso = matrix(0,nrow=nrow(test_Z),ncol=ncol(test_Z))\n",
    "    for (h in 1:ncol(mdl$W)){\n",
    "        \n",
    "        glmnet.mdl.cv <- cv.glmnet(x=train_X,y=Z_hat_tca[[h]],standardize=FALSE,alpha=1,nfolds=5)\n",
    "        rmse_lasso[h] = sqrt(glmnet.mdl.cv$cvm[glmnet.mdl.cv$lambda == glmnet.mdl.cv$lambda.min])\n",
    "        glmnet.mdl <- glmnet(x=train_X,y=Z_hat_tca[[h]],standardize=FALSE,alpha=1,lambda=glmnet.mdl.cv$lambda.min)\n",
    "        dev_ratio[h] <- glmnet.mdl$dev.ratio\n",
    "        beta.full <- as.numeric(glmnet.mdl$beta)\n",
    "        beta_full_cell[h,] <- beta.full\n",
    "        predictors <- colnames(train_X)[which(beta.full != 0)]\n",
    "        beta_lasso <- as.matrix(c(glmnet.mdl$a0,as.matrix(glmnet.mdl$beta[predictors,])))\n",
    "        Z_hat_train_lasso <- cbind(numeric(nrow(train_X))+1,train_X[,predictors]) %*% beta_lasso\n",
    "        Z_hat_test_lasso[,h] <- cbind(numeric(nrow(test_X))+1,test_X[,predictors]) %*% beta_lasso\n",
    "        Z_hat_test_lasso[,h] = signif(Z_hat_test_lasso[,h], digits = 5)\n",
    "        if(sum(beta_lasso)==0 | sd(Z_hat_train_lasso)==0){\n",
    "            # model 2 lasso forces SNPs effect to become zero.\n",
    "            # If there were no predictors, then cor would be NaN. Which also means\n",
    "            # no snps is correlated with Z. cor just set to 0.\n",
    "            corrs[h] = 0\n",
    "            corrs.real[h] = 0\n",
    "            corrs.beta[h] = 0\n",
    "            corrs.test.real[h] = 0\n",
    "        }else{\n",
    "            corrs[h] <- cor(t(Z_hat_tca[[h]]),Z_hat_train_lasso)\n",
    "            corrs.real[h] <- cor(train_Z[,h],Z_hat_train_lasso)\n",
    "            corrs.beta[h] <- cor(beta[h,],beta.full)\n",
    "            corrs.test.real[h] <- cor(test_Z[,h],Z_hat_test_lasso[,h])\n",
    "        }\n",
    "    }\n",
    "    # Precision & Recall for lasso\n",
    "    Binary_True=(beta!=0)\n",
    "    Binary_Pred_Lasso=(beta_full_cell!=0)\n",
    "    TP = do.call(rbind,lapply(1:cell_type,function(x) sum(as.integer(Binary_True[x,]&Binary_Pred_Lasso[x,]))))\n",
    "    FP = do.call(rbind,lapply(1:cell_type,function(x) sum(as.integer((!Binary_True[x,])&Binary_Pred_Lasso[x,]))))\n",
    "    FN = do.call(rbind,lapply(1:cell_type,function(x) sum(as.integer(Binary_True[x,]&(!Binary_Pred_Lasso[x,])))))\n",
    "    precision_lasso = TP/(TP+FP)\n",
    "    recall_lasso = TP/(TP+FN)\n",
    "    \n",
    "    return(list('Z_hat_test_lasso'=Z_hat_test_lasso,'Z_hat_tca'=Z_hat_tca,'Z_hat_cor_tca'=Z_hat_cor_tca,'beta_hat_cor_tca'=beta_hat_cor_tca,'gamma_hat_cor_tca'=gamma_hat_cor_tca,\n",
    "               'Z_hat_train_second_eq_cor_tca'=Z_hat_train_second_eq_cor_tca,'Z_hat_test_second_eq_cor_tca'=Z_hat_test_second_eq_cor_tca,\n",
    "               'G_hat_train_cor_bulk'=G_hat_train_cor_bulk,'G_hat_test_cor_bulk'=G_hat_test_cor_bulk,\n",
    "               'cor_lasso_tca'=corrs,'cor_lasso_real_train'=corrs.real,'cor_lasso_real_test'=corrs.test.real,\n",
    "               'cor_beta_lasso_real'=corrs.beta,'beta_hat_lasso'=beta_full_cell,'dev_rat'=dev_ratio,\n",
    "               'precision_lasso'=precision_lasso,'recall_lasso'=recall_lasso,'rmse_lasso'=rmse_lasso,'rmse_tca'=rmse_tca))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T21:59:33.232393Z",
     "start_time": "2019-09-03T21:59:33.144Z"
    }
   },
   "outputs": [],
   "source": [
    "generate_params <- function(cell_her=TRUE,gene_cor=FALSE,bulk_her=FALSE,seed=1,N=5000,M=100,herr_arr=c(),D=20,pslab=0.5,her=0.05,her_bulk=0.8){\n",
    "    set.seed(1)\n",
    "    if(length(herr_arr)!=0){\n",
    "        M = length(herr_arr)\n",
    "    }else{\n",
    "        M = M\n",
    "    }\n",
    "    K = 4\n",
    "    N = N\n",
    "    pc_num = 2\n",
    "    # cis_snps_nums = floor(runif(M, min = 150, max = 400))\n",
    "    D = D\n",
    "    cis_snps_nums = array(D,M)\n",
    "    sigma_g = 0.01\n",
    "    sigma_z = 0.1\n",
    "    \n",
    "    if(gene_cor){ # varing genetic correlation across genes\n",
    "        corr_seq = seq(from=0,to=pslab,length.out=M)\n",
    "    }else{\n",
    "        corr_seq = seq(from=0,to=0,length.out=M)\n",
    "    }\n",
    "    pslab = matrix(pslab,nrow=M,ncol=K)\n",
    "    corr_matrix = lapply(1:M,function(x) matrix(corr_seq[x],nrow=K,ncol=K))\n",
    "    for(i in 1:M){\n",
    "        diag(corr_matrix[[i]])=1\n",
    "    }\n",
    "    \n",
    "    if(cell_her){ # varing heribility across genes\n",
    "        heritibility_cell_specific = do.call(rbind,lapply(1:K,function(x) herr_arr))\n",
    "    }else{\n",
    "        heritibility_cell_specific = matrix(her,nrow=K,ncol=M)\n",
    "    }\n",
    "    \n",
    "    if(bulk_her){ # varing bulk heritibility across genes\n",
    "        heritibility_bulk = seq(from=0.1,to=her_bulk,length.out=M)\n",
    "    }else{\n",
    "        heritibility_bulk = array(her_bulk,M)\n",
    "    }\n",
    "        \n",
    "    MAF = lapply(1:length(cis_snps_nums),function(x) runif(cis_snps_nums[x],min=0.1,max=0.5))\n",
    "    id_name = sapply(1:N, function(x) paste('SAMPLE',toString(x),sep = ''))\n",
    "    cell_type_name=sapply(1:K, function(x) paste('Cell_type',x,sep = ''))\n",
    "    \n",
    "    ### generate C1 and C2\n",
    "    male = matrix(rbinom(N, 1, 0.5),nrow=N)\n",
    "    smoking = matrix(rbinom(N, 2, 0.2),nrow=N)\n",
    "    smoking = (smoking - min(smoking))/(max(smoking))\n",
    "    age = matrix(sapply(sapply(rnorm(N,50,20), function(x) floor(x)), function(x)  if (x<0){x=20} else{x=x}),nrow=N) \n",
    "    age = (age-min(age))/(max(age)-min(age))\n",
    "    c1 = scale(cbind(male, smoking, age))\n",
    "    c2 = scale(matrix(rnorm(N*pc_num), nrow=N))\n",
    "                            \n",
    "    rownames(c1) = id_name\n",
    "    colnames(c1) = c('male','smoking','age')\n",
    "    rownames(c2) = id_name\n",
    "    colnames(c2) = sapply(1:pc_num, function(x) paste('PC',toString(x),sep = ''))\n",
    "                          \n",
    "    p1 = dim(c1)[2]\n",
    "    sigma_gamma = sqrt((1-heritibility_cell_specific[1,]-sigma_z^2)/p1)\n",
    "    var_beta = heritibility_cell_specific[1,]*(p1*sigma_gamma^2+sigma_z^2)/(1-heritibility_cell_specific[1,])/D\n",
    "    \n",
    "    # cell type proportion, from the real data estimated\n",
    "    W_alpha = c(26.553683792256,17.6621467979005,4.48671525658667,1.56874856517803)#,0.178797663350679)\n",
    "    #W_alpha = 50.4500920752719\n",
    "    W_xsi = c(0.526335685426257,0.350091468050216,0.088933737720289,0.0310950585152043,0.00354405028803339)\n",
    "    if(K <= length(W_alpha)){\n",
    "        W = rdirichlet(N, W_alpha[1:K])\n",
    "    }else{ #more cell type prop\n",
    "        W = rdirichlet(N, runif(K, 0, 1))\n",
    "    }\n",
    "    colnames(W) = cell_type_name\n",
    "    rownames(W) = id_name\n",
    "    \n",
    "    alpha = W_alpha\n",
    "    alpha_0 = sum(alpha)\n",
    "    alpha_tilde = alpha/alpha_0\n",
    "    m2_alpha = alpha_tilde %*% t(alpha_tilde)*alpha_0/(alpha_0+1)\n",
    "    diag(m2_alpha) = alpha_tilde*(1-alpha_tilde)/(alpha_0+1)+alpha_tilde^2\n",
    "    \n",
    "    return(list(\n",
    "        'M'=M,'K'=K,'N'=N,'pc_num'=pc_num,'D'=D,'id_name'=id_name,'cell_name'=cell_type_name,\n",
    "        'her'=heritibility_cell_specific,'her_bulk'=heritibility_bulk,'pslab'=pslab,\n",
    "        'MAF'=MAF,'c1'=c1,'c2'=c2,'W'=W,'sigma_gamma'=sigma_gamma,'sigma_g'=sigma_g,'m2_alpha'=m2_alpha,\n",
    "        'sigma_z'=sigma_z,'beta_cor'=corr_matrix,'var_beta'=var_beta,'W_alpha'=W_alpha[1:K]))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T21:59:33.760776Z",
     "start_time": "2019-09-03T21:59:33.670Z"
    }
   },
   "outputs": [],
   "source": [
    "one_gene <- function(par,g,seed=1){\n",
    "    set.seed(seed)\n",
    "    cell_type = par$K\n",
    "    n_snps = par$D\n",
    "    her = par$her[,g]\n",
    "    cor = par$beta_cor[[g]]\n",
    "    pslab = par$pslab[g,]\n",
    "    var_beta = array(par$var_beta[g],par$K)\n",
    "    \n",
    "    maf = par$MAF[[g]]\n",
    "    dummy_var = 10\n",
    "    \n",
    "    var_matrix=sqrt(var_beta%*%t(var_beta))\n",
    "    Sigma_beta = cor*var_matrix/(1-pslab)^2\n",
    "    diag(Sigma_beta) = diag(Sigma_beta) *(1-pslab)\n",
    "    beta_ = mvrnorm(n_snps,mu=rep(0,par$K),Sigma=Sigma_beta,tol=1e-4,empirical=TRUE)\n",
    "    rmask = do.call(rbind,lapply(1:cell_type,function(x) rbinom(n_snps,1,1-pslab[x])))\n",
    "    beta = t(beta_)*rmask\n",
    "\n",
    "    X = t(do.call(rbind,lapply(1:n_snps,function(x) rbinom(par$N,2,maf[x]))))\n",
    "    colnames(X) = sapply(1:n_snps, function(x) paste('G',g,'_SNP',toString(x),sep = ''))\n",
    "    rownames(X) = par$id_name\n",
    "    X = scale(X)\n",
    "    # enforce variance of samples by adjusting sd of beta\n",
    "    xbeta_var = n_snps*var_beta[1]\n",
    "    beta = do.call(rbind,\n",
    "                      lapply(1:cell_type,function(x) beta[x,]*(sqrt(xbeta_var)/sd(X %*% beta[x,]))))\n",
    "\n",
    "    mu_z = X%*%t(beta)\n",
    "    epsilon_z = array(rnorm(par$N*cell_type,mean=0,sd=par$sigma_z), c(par$N,cell_type))\n",
    "    epsilon_z = apply(epsilon_z,2,function(x) x*par$sigma_z/sd(x))\n",
    "    gamma_c1 = do.call(rbind,lapply(1:cell_type,function(x) rnorm(dim(par$c1)[2],mean=0,sd=par$sigma_gamma)))     # enforce variance of samples by adjusting sd of gamma_c1\n",
    "    # enforce the c1_gamma variance\n",
    "    c1gamma_var = dim(par$c1)[2]*(par$sigma_gamma[g])^2\n",
    "    gamma_c1 = do.call(rbind,\n",
    "                       lapply(1:cell_type,function(x) gamma_c1[x,]*(sqrt(c1gamma_var)/sd(par$c1 %*% gamma_c1[x,]))))\n",
    "    c1_gamma = par$c1 %*% t(gamma_c1)\n",
    "\n",
    "    Z = epsilon_z + mu_z + c1_gamma\n",
    "    rownames(Z) = par$id_name\n",
    "    colnames(Z) = par$cell_name\n",
    "    \n",
    "    ### Generate Gene expressions\n",
    "    G = rowSums((par$W)*Z)\n",
    "    epsilon_G = rnorm(par$N,mean=0,sd=par$sigma_g)\n",
    "    epsilon_G = epsilon_G*par$sigma_g/sd(epsilon_G)\n",
    "    \n",
    "    # use bulk level heritability to calculate sigma_gamma\n",
    "    bulk_her_nom = sum(hadamard.prod(par$m2_alpha,var(mu_z)))\n",
    "    bulk_her_c1_gamma = sum(hadamard.prod(par$m2_alpha,var(c1_gamma)))\n",
    "    bulk_her_epsilon_z = sum(hadamard.prod(par$m2_alpha,var(epsilon_z)))\n",
    "    bulk_her_z = sum(hadamard.prod(par$m2_alpha,var(Z)))\n",
    "    \n",
    "    bulk_her_denom = sum(hadamard.prod(par$m2_alpha,var(Z)))+var(epsilon_G)\n",
    "    bulk_her_val = par$her_bulk[g]*her[1]\n",
    "    sd_gamma_c2 = sqrt((bulk_her_nom/bulk_her_val-bulk_her_denom)/par$pc_num)\n",
    "    \n",
    "    gamma_c2 = matrix(rnorm(par$pc_num,mean=0,sd=sd_gamma_c2),nrow=par$pc_num,ncol=1)\n",
    "    c2gamma_var = par$pc_num * (sd_gamma_c2)^2\n",
    "    gamma_c2 = gamma_c2*(sqrt(c2gamma_var)/sd(par$c2%*%gamma_c2))\n",
    "    c2_gamma = par$c2 %*% gamma_c2\n",
    "    \n",
    "    G = t(G + epsilon_G + c2_gamma)\n",
    "    real_denom = var(t(G))\n",
    "    colnames(G) <- par$id_name\n",
    "    rownames(G) <- paste('gene',g,sep='_')\n",
    "    G = as.data.frame(G)\n",
    "    \n",
    "    real_bulk_her = bulk_her_nom/(bulk_her_denom+var(c2_gamma))\n",
    "                              \n",
    "    return(list('X'=X,'beta'=beta,'gamma_c1'=t(gamma_c1),'bulk_her'=real_bulk_her,\n",
    "               'gamma_c2'=gamma_c2,'Z'=Z,'G'=G,'epsilon_z'=epsilon_z))\n",
    "}\n",
    "#data = one_gene(par=params,1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Power plot\n",
    "more sample, power should grow as well\n",
    "Hold all other paramters the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T22:00:38.716392Z",
     "start_time": "2019-09-03T22:00:38.597Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "lmp <- function (modelobject) {\n",
    "    if (class(modelobject) != \"lm\") stop(\"Not an object of class 'lm' \")\n",
    "    f <- summary(modelobject)$fstatistic\n",
    "    p <- pf(f[1],f[2],f[3],lower.tail=F)\n",
    "    attributes(p) <- NULL\n",
    "    return(p)\n",
    "}\n",
    "summary_statistics_z_hat_test_lasso <- function(mdl,train_X,test_X,train_c1,test_c1,\n",
    "                               train_G,test_G,train_Z,test_Z,beta,\n",
    "                               gamma_c1,gamma_c2,model=1){\n",
    "    \n",
    "    Z_hat_tca = tensor(train_G,mdl,log_file=NULL,verbose=FALSE,debug=FALSE)\n",
    "    Z_hat_test_lasso = matrix(0,nrow=nrow(test_Z),ncol=ncol(test_Z))\n",
    "    beta_full_cell = matrix(0,nrow=ncol(mdl$W),ncol=dim(train_X)[2])\n",
    "    for (h in 1:ncol(mdl$W)){\n",
    "        glmnet.mdl.cv <- cv.glmnet(x=train_X,y=Z_hat_tca[[h]],standardize=FALSE,alpha=1,nfolds=5)\n",
    "        glmnet.mdl <- glmnet(x=train_X,y=Z_hat_tca[[h]],standardize=FALSE,alpha=1,lambda=glmnet.mdl.cv$lambda.min)\n",
    "        beta.full <- as.numeric(glmnet.mdl$beta)\n",
    "        beta_full_cell[h,] <- beta.full\n",
    "        predictors <- colnames(train_X)[which(beta.full != 0)]\n",
    "        beta_lasso <- as.matrix(c(glmnet.mdl$a0,as.matrix(glmnet.mdl$beta[predictors,])))\n",
    "        Z_hat_test_lasso[,h] <- cbind(numeric(nrow(test_X))+1,test_X[,predictors]) %*% beta_lasso\n",
    "        Z_hat_test_lasso[,h] = signif(Z_hat_test_lasso[,h], digits = 5)\n",
    "    }\n",
    "    \n",
    "    return(list('Z_hat_test_lasso'=Z_hat_test_lasso,'beta_hat_lasso'=beta_full_cell))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T22:15:30.201648Z",
     "start_time": "2019-09-03T22:01:45.415Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#params = generate_params(cont_her='other',gene_cor=FALSE,seed=1,N=105000,M=1000,D=20,pslab=0.5,her=0.05)\n",
    "params_sample_check = generate_params(cell_her=FALSE,gene_cor=FALSE,bulk_her=FALSE,\n",
    "                         seed=1,N=105000,M=1000,\n",
    "                         D=20,pslab=0.5,her=0.05,\n",
    "                         her_bulk=0.8)\n",
    "data_sample_check = lapply(1:params_sample_check$M,function(x) one_gene(par=params_sample_check,x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_size_arr = c(100,200,300,400,500)\n",
    "GWAS_size = 100000\n",
    "li_sample_check_power = list()\n",
    "p_value_matrix = array(0,c(length(sample_size_arr),length(data_sample_check),params_sample_check$K))\n",
    "for(sample_size_i in 1:length(sample_size_arr)){\n",
    "    sample_size = sample_size_arr[sample_size_i]\n",
    "    for(g in 1:length(data_sample_check)){\n",
    "        train_X = (data_sample_check[[g]]$X)[1:sample_size,]\n",
    "        test_X = data_sample_check[[g]]$X[(params_sample_check$N-GWAS_size+1):params_sample_check$N,]\n",
    "        train_W = params_sample_check$W[1:sample_size,]\n",
    "        test_W = params_sample_check$W[(params_sample_check$N-GWAS_size+1):params_sample_check$N,]\n",
    "        train_G = data_sample_check[[g]]$G[,1:sample_size]\n",
    "        test_G = data_sample_check[[g]]$G[,(params_sample_check$N-GWAS_size+1):params_sample_check$N]\n",
    "        train_c1 = params_sample_check$c1[1:sample_size,]\n",
    "        test_c1 = params_sample_check$c1[(params_sample_check$N-GWAS_size+1):params_sample_check$N,]\n",
    "        train_c2 = params_sample_check$c2[1:sample_size,]\n",
    "        test_c2 = params_sample_check$c2[(params_sample_check$N-GWAS_size+1):params_sample_check$N,]\n",
    "        train_Z = data_sample_check[[g]]$Z[1:sample_size,]\n",
    "        test_Z = data_sample_check[[g]]$Z[(params_sample_check$N-GWAS_size+1):params_sample_check$N,]\n",
    "        beta = data_sample_check[[g]]$beta\n",
    "        gamma_c1 = data_sample_check[[g]]$gamma_c1\n",
    "        gamma_c2 = data_sample_check[[g]]$gamma_c2\n",
    "        tca.mdl1 = tca(X=train_G,W=train_W,C1=cbind(train_X,train_c1),verbose=FALSE)\n",
    "        tca.mdl1.summary = summary_statistics_z_hat_test_lasso(mdl=tca.mdl1,train_X=train_X,test_X=test_X,\n",
    "                                                  train_c1=train_c1,test_c1=test_c1,train_G=train_G,\n",
    "                                                  test_G=test_G,train_Z=train_Z,test_Z=test_Z,\n",
    "                                                  beta=beta,gamma_c1=gamma_c1,gamma_c2=gamma_c2,model=1)\n",
    "        li_sample_check_power[[paste(g,'.',sample_size,sep='')]] = tca.mdl1.summary\n",
    "        r_testz = dim(test_Z)[1]\n",
    "        c_testz = dim(test_Z)[2]\n",
    "        Y = test_Z + matrix(rnorm(r_testz*c_testz,mean=0,sd=sqrt(499)),nrow=r_testz)\n",
    "        for(cell_i in 1:params_sample_check$K){\n",
    "            if(length(unique(tca.mdl1.summary$Z_hat_test_lasso[,cell_i]))==1){\n",
    "                p_value_matrix[sample_size_i,g,cell_i] = 1\n",
    "                next #If lasso predict everything to be the same. p value breaks\n",
    "            }\n",
    "            lm.model <- lm(Y[,cell_i]~tca.mdl1.summary$Z_hat_test_lasso[,cell_i])\n",
    "            p_value_matrix[sample_size_i,g,cell_i] = lmp(lm.model)            \n",
    "        }\n",
    "    }\n",
    "    print(paste('sample size',sample_size,'done'))\n",
    "}\n",
    "#save(li_sample_check_power,sample_size_arr,params,data, file = \"power_check.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T20:56:53.814368Z",
     "start_time": "2019-09-03T20:56:53.761Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# # For changing the phenotype heritability value 499/1999.\n",
    "# # No need to run this if 499 up in the above cell is correct\n",
    "# p_value_matrix = array(0,c(length(sample_size_arr),length(data),params$K))\n",
    "# for(sample_size_i in 1:length(sample_size_arr)){\n",
    "#     sample_size = sample_size_arr[sample_size_i]\n",
    "#     for(g in 1:length(data)){\n",
    "#         test_Z = data[[g]]$Z[(params$N-GWAS_size+1):params$N,]\n",
    "#         r_testz = dim(test_Z)[1]\n",
    "#         c_testz = dim(test_Z)[2]\n",
    "#         Y = test_Z + matrix(rnorm(r_testz*c_testz,mean=0,sd=sqrt(499)),nrow=r_testz)\n",
    "#         Z_hat = li_sample_check_power[[paste(g,'.',sample_size,sep='')]]$Z_hat_test_lasso\n",
    "#         for(cell_i in 1:params$K){\n",
    "#             if(length(unique(Z_hat[,cell_i]))==1){\n",
    "#                 p_value_matrix[sample_size_i,g,cell_i] = 1\n",
    "#                 next #If lasso predict everything to be the same. p value breaks\n",
    "#             }\n",
    "#             lm.model <- lm(Y[,cell_i]~Z_hat[,cell_i])\n",
    "#             p_value_matrix[sample_size_i,g,cell_i] = lmp(lm.model)            \n",
    "#         }\n",
    "#     }\n",
    "#     print(paste('sample size',sample_size,'done'))\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mean_weight = signif(apply(params_sample_check$W,2,mean),digits=3)\n",
    "par(mfrow=c(2,2),mar=c(4, 3.8, 4, 2) + 0.1)\n",
    "opts=params_sample_check$cell_name\n",
    "for(i in 1:params_sample_check$K){\n",
    "    temp = data.frame(t(p_value_matrix[,,i]))\n",
    "    colnames(temp) = sample_size_arr\n",
    "    heading = paste(opts[i],'mean weight',mean_weight[i],sep=' ')\n",
    "    boxplot(temp,xlab='Training Sample Size',ylab='p value',cex.lab=1.2,ylim=c(0,1),yaxs='r',type='o',col='dodgerblue')\n",
    "    title(main=heading,font.main=2,cex.main=1,line=1,cex.lab=1)\n",
    "}\n",
    "mtext(\"P value on test data set VS sample size\",side=3,adj=3.5,cex=1,line=23.5,font=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_threshold = 0.05/10000/4\n",
    "power_matrix = matrix(0,nrow=params_sample_check$K,ncol=length(sample_size_arr))\n",
    "for(cell_i in 1:params_sample_check$K){\n",
    "    power_matrix[cell_i,] = t(do.call(rbind,lapply(1:length(sample_size_arr),\n",
    "                                                 function(x) sum(as.integer(p_value_matrix[x,,cell_i]<p_threshold)))))\n",
    "}\n",
    "power_matrix = power_matrix/params_sample_check$M\n",
    "\n",
    "mean_weight = signif(apply(params_sample_check$W,2,mean),digits=3)\n",
    "par(mfrow=c(2,2),mar=c(4, 3.8, 4, 2) + 0.1)\n",
    "opts=params_sample_check$cell_name\n",
    "for(i in 1:params_sample_check$K){\n",
    "    heading = paste(opts[i],'mean weight',mean_weight[i],sep=' ')\n",
    "    plot(sample_size_arr,power_matrix[i,],xlab='Sample Size',\n",
    "         ylab='power',cex.lab=1.2,ylim=c(0,0.1),yaxs='r',type='o',col='dodgerblue')\n",
    "    title(main=heading,font.main=2,cex.main=1,line=1,cex.lab=1)\n",
    "}\n",
    "mtext(\"Significant P value on test data set VS sample size\",side=3,adj=1.7,cex=1,line=23.5,font=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "lot.title <- \"Power of cell-specific expression imputation\"\n",
    "x.label <- \"Training sample size\"\n",
    "y.label <- \"Power\"\n",
    "legend.labels <- c('High abundance', 'Intermediate abundance', 'Low abundance')\n",
    "\n",
    "par(pin=c(3.5,3.5)) # plot size\n",
    "cell.type.colors <- c('dodgerblue', 'darkorange', 'seagreen3')\n",
    "marker.style = 20 \n",
    "smoothing.param = 0.35 # value in (0,1], 0 is original, 1 plots a straight line\n",
    "\n",
    "#' End plot parameters\n",
    "\n",
    "power.plot.data.path <- \"/u/scratch/b/bjew/power.plot.rds\"\n",
    "\n",
    "power.plot.data <- readRDS(power.plot.data.path)\n",
    "power.matrix <- power.plot.data$power.matrix # Entry {i,j} corresponds to power for cell type i at sample size j\n",
    "sample.sizes <- power.plot.data$sample.sizes # Array of sample sizes\n",
    "plot(NULL, xlim=c(min(sample.sizes),max(sample.sizes)), ylim=c(0,1), ylab=y.label, xlab=x.label)\n",
    "for (i in 1:nrow(power.matrix)){\n",
    "    smoothingSpline = smooth.spline(sample.sizes, power.matrix[i,],\n",
    "                                    spar=smoothing.param)\n",
    "    lines(smoothingSpline, col=cell.type.colors[i])\n",
    "    points(smoothingSpline, col=cell.type.colors[i], pch=marker.style)\n",
    "}\n",
    "\n",
    "# font.main (1=regular, 2=bold), cex.main (size)\n",
    "title(plot.title,font.main=2,cex.main=1)\n",
    "legend(legend=legend.labels,col=cell.type.colors,\n",
    "       x='topleft',inset=0.03,lwd=2,cex=0.6,horiz=FALSE, bty='n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sample Size Check\n",
    "Correlation square is just r square, the proportion of variance explained"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T21:48:58.524121Z",
     "start_time": "2019-09-03T21:48:58.426Z"
    }
   },
   "outputs": [],
   "source": [
    "summary_statistics_corr <- function(mdl,train_X,test_X,train_c1,test_c1,\n",
    "                               train_G,test_G,train_Z,test_Z,beta,\n",
    "                               gamma_c1,gamma_c2,model=1){\n",
    "    # TCA estimation & parameters\n",
    "    Z_hat_tca = tensor(train_G,mdl,log_file=NULL,verbose=FALSE,debug=FALSE)\n",
    "    # cell type specific lasso\n",
    "    corrs.test.real = numeric(ncol(mdl$W))\n",
    "    Z_hat_test_lasso = matrix(0,nrow=nrow(test_Z),ncol=ncol(test_Z))\n",
    "    for (h in 1:ncol(mdl$W)){\n",
    "        glmnet.mdl.cv <- cv.glmnet(x=train_X,y=Z_hat_tca[[h]],standardize=FALSE,alpha=1,nfolds=5)\n",
    "        glmnet.mdl <- glmnet(x=train_X,y=Z_hat_tca[[h]],standardize=FALSE,alpha=1,lambda=glmnet.mdl.cv$lambda.min)\n",
    "        beta.full <- as.numeric(glmnet.mdl$beta)\n",
    "        predictors <- colnames(train_X)[which(beta.full != 0)]\n",
    "        beta_lasso <- as.matrix(c(glmnet.mdl$a0,as.matrix(glmnet.mdl$beta[predictors,])))\n",
    "        Z_hat_train_lasso <- cbind(numeric(nrow(train_X))+1,train_X[,predictors]) %*% beta_lasso\n",
    "        Z_hat_test_lasso[,h] <- cbind(numeric(nrow(test_X))+1,test_X[,predictors]) %*% beta_lasso\n",
    "        Z_hat_test_lasso[,h] = signif(Z_hat_test_lasso[,h], digits = 5)\n",
    "        if(sum(beta_lasso)==0 | sd(Z_hat_train_lasso)==0){\n",
    "            # model 2 lasso forces SNPs effect to become zero.\n",
    "            # If there were no predictors, then cor would be NaN. Which also means\n",
    "            # no snps is correlated with Z. cor just set to 0.\n",
    "            corrs.test.real[h] = 0\n",
    "        }else{\n",
    "            corrs.test.real[h] <- cor(test_Z[,h],Z_hat_test_lasso[,h])\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    return(list('cor_lasso_real_test'=corrs.test.real))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T21:48:59.817458Z",
     "start_time": "2019-09-03T21:48:59.394Z"
    }
   },
   "outputs": [],
   "source": [
    "herr_arr = c(0.01,0.05,0.1,0.2)\n",
    "#params = generate_params(cont_her='other',gene_cor=FALSE,seed=1,N=105000,M=100,D=20,pslab=0.5,her=0.05)\n",
    "params_her_check = generate_params(cell_her=TRUE,gene_cor=FALSE,bulk_her=FALSE,\n",
    "                         seed=1,N=5000,M=4,herr_arr=herr_arr,\n",
    "                         D=20,pslab=0.5,her=0.05,\n",
    "                         her_bulk=0.8)\n",
    "data_heritability_check = lapply(1:params_her_check$M,function(x) one_gene(par=params_her_check,x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T21:51:44.972791Z",
     "start_time": "2019-09-03T21:50:22.343Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"gene_1\"\n",
      "[1] \"gene_2\"\n",
      "[1] \"gene_3\"\n",
      "[1] \"gene_4\"\n"
     ]
    }
   ],
   "source": [
    "# do TCA-TWAS for every gene, every sample size\n",
    "prop = 0.7\n",
    "sample_size_arr = c(1500,2000,2500,3000,3500,4000,4500,5000)\n",
    "t_prop = floor(prop*params_her_check$N)\n",
    "li_her = list()\n",
    "for(g in 1:length(data_heritability_check)){\n",
    "    print(paste('gene',g,sep='_'))\n",
    "    for(sample_size in sample_size_arr){\n",
    "        t_prop = floor(prop*sample_size)\n",
    "        train_X = (data_heritability_check[[g]]$X)[1:t_prop,]\n",
    "        test_X = data_heritability_check[[g]]$X[(t_prop+1):sample_size,]\n",
    "        train_W = params_her_check$W[1:t_prop,]\n",
    "        test_W = params_her_check$W[(t_prop+1):sample_size,]\n",
    "        train_G = data_heritability_check[[g]]$G[,1:t_prop]\n",
    "        test_G = data_heritability_check[[g]]$G[,(t_prop+1):sample_size]\n",
    "        train_c1 = params_her_check$c1[1:t_prop,]\n",
    "        test_c1 = params_her_check$c1[(t_prop+1):sample_size,]\n",
    "        train_c2 = params_her_check$c2[1:t_prop,]\n",
    "        test_c2 = params_her_check$c2[(t_prop+1):sample_size,]\n",
    "        train_Z = data_heritability_check[[g]]$Z[1:t_prop,]\n",
    "        test_Z = data_heritability_check[[g]]$Z[(t_prop+1):sample_size,]\n",
    "        beta = data_heritability_check[[g]]$beta\n",
    "        gamma_c1 = data_heritability_check[[g]]$gamma_c1\n",
    "        gamma_c2 = data_heritability_check[[g]]$gamma_c2\n",
    "        tca.mdl1 = tca(X=train_G,W=train_W,C1=cbind(train_X,train_c1),verbose=FALSE)\n",
    "        tca.mdl1.summary = summary_statistics_corr(mdl=tca.mdl1,train_X=train_X,test_X=test_X,\n",
    "                                              train_c1=train_c1,test_c1=test_c1,train_G=train_G,\n",
    "                                              test_G=test_G,train_Z=train_Z,test_Z=test_Z,\n",
    "                                              beta=beta,gamma_c1=gamma_c1,gamma_c2=gamma_c2,model=1)\n",
    "        li_her[[paste(g,'.',sample_size,sep='')]] = list('s1'=tca.mdl1.summary)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-03T22:21:20.728689Z",
     "start_time": "2019-09-03T22:21:20.401Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"pred_of_cis-regulated_cell-specific_expression_from_bulk_data.pdf\")\n",
    "cell_type_weight_mean = signif(apply(params_her_check$W,2,mean),3)\n",
    "legend = lapply(1:4,function(x) paste('cell',x,'_w_',cell_type_weight_mean[x],sep=''))\n",
    "par(mfrow=c(2,2))\n",
    "opts=paste('Heritability = ',params_her_check$her[1,])\n",
    "for(i in 1:length(data_heritability_check)){\n",
    "    cor_sample_size = do.call(rbind,lapply(1:length(sample_size_arr),\n",
    "                                                    function(x) li_her[[paste(i,'.',sample_size_arr[x],sep='')]]$s1$cor_lasso_real_test))\n",
    "    plot(sample_size_arr,cor_sample_size[,1]^2,xlab='Sample Size',\n",
    "         ylab=parse(text='Correlation (r^2)'),cex.lab=0.9,ylim=c(0,0.2),yaxs='r',type='o',col='dodgerblue')\n",
    "    title(main=opts[i],font.main=2,cex.main=1,line=1,cex.lab=1)\n",
    "    lines(sample_size_arr,cor_sample_size[,2]^2,col='black',type='o')\n",
    "    lines(sample_size_arr,cor_sample_size[,3]^2,col='purple',type='o')\n",
    "    lines(sample_size_arr,cor_sample_size[,4]^2,col='blue',type='o')\n",
    "}\n",
    "mtext(\"Prediction of cis-regulated cell-specfic expression from bulk data\",side=3,adj=1.1,cex=1,line=24,font=2)\n",
    "par(mfrow=c(2,1),mar=c(0,0,0,0))\n",
    "plot(1,type='n',axes=FALSE,xlab='',ylab='')\n",
    "legend(x='top',inset=0,legend=legend,\n",
    "       col=c('dodgerblue','black','purple','blue'),lwd=2,cex=0.7,horiz=TRUE)\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
